File: animax/src/resource/resource_loader/ios/AnimaXBase64Loader.m
// Copyright 2024 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.

#import "AnimaXBase64Loader.h"

@implementation AnimaXBase64Loader
- (void)handleRequest:(AnimaXLoaderRequest*)request
           completion:(AnimaXLoaderCompletionHandler)completion {
  AnimaXLoaderResponse* response = nil;
  do {
    if (nil == request || nil == request.url || [request.url length] == 0) {
      response = [AnimaXLoaderResponse responseWithErrorMessage:@"Invalid AnimaXLoaderRequest"];
      break;
    }
    NSURL* url = [NSURL URLWithString:request.url];
    if (nil == url) {
      response = [AnimaXLoaderResponse responseWithErrorMessage:@"Invalid URL"];
      break;
    }
    NSError* error = nil;
    NSData* data = [NSData dataWithContentsOfURL:url options:0 error:&error];
    if (nil != error) {
      response = [AnimaXLoaderResponse responseWithError:error];
      break;
    }

    if (nil == data || [data length] == 0) {
      response =
          [AnimaXLoaderResponse responseWithErrorMessage:@"Cannot convert base64 URL to NSData."];
      break;
    }

    response = [AnimaXLoaderResponse responseWithPayload:data];
  } while (0);

  completion(response);
}

- (AnimaXLoaderScheme)getScheme {
  return AnimaXLoaderSchemeDataURL;
}
@end

File: platform/android/animax_android/src/main/java/com/lynx/animax/loader/AnimaXBase64Loader.java
// Copyright 2024 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.
package com.lynx.animax.loader;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.ColorSpace;
import android.os.Build;
import android.util.Base64;

public class AnimaXBase64Loader implements IAnimaXLoader {
  private static final String BASE64_PREFIX = "base64,";

  private static String getDataUrlBase64Content(String dataUrl) {
    return dataUrl.substring(dataUrl.indexOf(BASE64_PREFIX) + BASE64_PREFIX.length());
  }

  private Bitmap decodeBitmap(byte[] data) {
    BitmapFactory.Options options = new BitmapFactory.Options();
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
      options.inPreferredColorSpace = ColorSpace.get(ColorSpace.Named.SRGB);
    }
    return BitmapFactory.decodeByteArray(data, 0, data.length, options);
  }

  @Override
  public void load(IAnimaXLoaderRequest request, IAnimaXLoaderCompletionHandler completionHandler) {
    AnimaXLoaderResponse<?> response = null;
    try {
      String base64DataUrl = request.getUri();
      byte[] data = Base64.decode(getDataUrlBase64Content(base64DataUrl), Base64.DEFAULT);

      if (request.getImageInfo() != null) {
        Bitmap bitmap = decodeBitmap(data);
        response =
            AnimaXLoaderResponse.createBitmapResponse(new SimpleCloseableBitmapReference(bitmap));
      } else {
        response = AnimaXLoaderResponse.createByteArrayResponse(data);
      }
    } catch (Exception e) {
      response = AnimaXLoaderResponse.createErrorResponse(e);
    } finally {
      completionHandler.onComplete(response);
    }
  }

  @Override
  public AnimaXLoaderScheme getScheme() {
    return AnimaXLoaderScheme.DATA_URL;
  }
}
